Exercise/collision check

HUmutOzdemir opened the pull request.

alezana commented:
@HUmutOzdemir 
Some comments:
- [ ] I don't really like the 
> "But you are not allowed to use modules that implements collision check directly such as `shapely`."
hard for us to check, and I'm not 100% sure how convenient it would be for them...
- [ ] I would add some function description in CollisonPrimitives class, also some methods have the same signature which seems a bit odd, in teh second one is the polygon supposed to be an aabb rectangle? then I would not pass a polygon but an aabb rectangle (fixed)
- [ ] a polygon and a path are essentially the same thing according to the current definition. I feel it is missin a __post_init__ method in the polygon making sure that it is actually closed (first==last point) or/and that it closes it by default
- [ ] Line I am not sure it is the most appropriate term.... maybe segment is more appropriate? I would check against the "standard" geometry libraries
- [ ] Unclear how thje performance metrics are computed... not in the doc atm, please add them 
- [ ] For the second part maybe we can tailor the environment to actually shine with the method you are asking them to implement (number of obstacles up in rtree, number of points along the path up in the safety certificates one,....)

HUmutOzdemir commented:
@alezana 

1. I found a way to list the modules imported in a file (by using `globals()`). However, I can not check wheter a function is using `shapely` or not. I am open to any suggestion. 

2. There must have been a misunderstanding. In fact, the purpose of that function is also to determine whether there is an collision between a polygon and a line segment. But the idea is using AABB to increase performance by skipping that point if the point is outside of AABB. This is an example implementation of this function.

```python
    @staticmethod
    def polygon_line_collision_aabb(p: Polygon, line: Line) -> bool:

        aabb = CollisionPrimitives.convert_poly_to_AABB(p)
        sampled_points = CollisionPrimitives.sample_line(line)

        for point in sampled_points:

            if aabb.p_min.x > point.x or aabb.p_min.y > point.y:
                continue

            if aabb.p_max.x < point.x or aabb.p_max.y < point.y:
                continue

            if CollisionPrimitives.polygon_point_collision(p, point):
                return True

        return False
```

3. Like the polygon definition in shapely, our Polygon class closes the loop by default. However, there is not a connection between first and last point for the Path class. I added additional note to the documentation to make this difference clear.

```
    Please note that the definition of `Path` and `Polygons` are similar. `Polygon` class connects the first and last vertices by default. However, there is not any connection between first and last waypoints in `Path` objects. 
```

4. I checked SymPy, and they are using `Segment`. I changed the name of the class as `Segment`. If you will find more appropriate name, I may change it again.

5. I added new part to documentation as Evaluation.

6. I don't understand how I can shine with the method they are asked to implement. Could you give me more details, examples etc.?
